crates/data-model/src/oauth2/client.rs

```diff
@@ -175,3 +175,66 @@ impl Client {
         ]
     }
 }
+
+/// The hosts that match the loopback interface.
+const LOCAL_HOSTS: &[&str] = &["localhost", "127.0.0.1", "[::1]"];
+
+/// Whether the given URI matches one of the registered URIs.
+///
+/// If the URI host is one if `localhost`, `127.0.0.1` or `[::1]`, any port is
+/// accepted.
+fn uri_matches_one_of(uri: &Url, registered_uris: &[Url]) -> bool {
+    if LOCAL_HOSTS.contains(&uri.host_str().unwrap_or_default()) {
+        for registered_uri in registered_uris {
+            if registered_uri.scheme() == uri.scheme()
+                && registered_uri.host() == uri.host()
+                && registered_uri.path() == uri.path()
+            {
+                return true;
+            }
+        }
+
+        false
+    } else {
+        registered_uris.contains(uri)
+    }
```

Comment:
I'm tempted to just try to strip the port on the URI if it's localhost? That would require that the registered URI don't have a port in them though... which might be fine? I can't really find a good source in RFCs that mixes dynamic registration and native apps with localhost/127.0.0.1 as the redirector :(

Something like that?
```suggestion
    if LOCAL_HOSTS.contains(&uri.host_str().unwrap_or_default()) {
        let mut uri = uri.clone();
        // Try matching without the port first
        if url.set_port(None).is_ok() {
            if registered_uris.contains(uri) {
                return true;
            }
        }
    } 
    
    registered_uris.contains(uri)
```

Reply:
Yes, the clearest I got was from [OAuth 2.1](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-08#name-registration-requirements):

> Authorization servers MUST reject authorization requests that specify a redirect URI that doesn't exactly match one that was registered, with an exception for loopback redirects, where an exact match is required except for the port URI component.

Which to me means we should check everything but the port. But if you prefer it your way, it's fine by me.